### ISSUE TITLE
[D8] Add ability to create configurable layout templates

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#3755]
We should have flexible layouts in core.

Currently to create a new layout, you have to choose a template. Templates are provided in code, in a package with an info file and a tpl file, similar to modules or themes. So to create a new template requires some coding. 

This PR allows creating a new template in the UI, by adding an unlimited number of rows, with each row optionally divided into preset columns. Each column a layout region. 

There is another issue somewhere, but I cannot find it.

MVP (BLOCKER) TODOs: 
* [x] Rename `Columns` to `Regions` everywhere (for consistent terminology)
* [x] Change `Configure` link in drop button to `Configure regions`
* [x] Change `Edit` from drop-button; use `Configure name` instead (edit means db storage is used)
* [x] Hide & automate row names (b/c they should not appear in the UI anywhere else)
* [x] Region names should start from 1 and not zero (Numbers should not reset per row.)
* [x] Region names should default to "Region 1" (region_1), "Region 2" (region_2), "Region 3" (region_3)
* [x] Make rows draggable to re-order
* [x] Add a unique icon to use for flexible layouts (can we change the color, maybe?)
* [x] Basic automated testing

ACTIONABLE TODOs:
* [x] Add ability to re-order rows
* [x] A single 'Add new row' button at the bottom
* [x] Dialog title for adding a row should be `Add new row` or similar
* [x] Move region names out of fieldset or open by default (b/c they **do** appear in the manage blocks UI)
* [x] Field label `Change column style` should be `Number of regions` (no style setting here)
* [x] Remove the `Select number of regions` option from the `Number of regions` select list; set the default to 1, and also select the first available option from the region widths.
* [x] Change text `Column style` to `Column widths` (no style setting here)
* [x] Change ratios to approximate percents in column width options 
* [x] Change text `Selected column style` to `Selected column widths` (no style setting here)
* [x] Default drop-button action for flexible layout should be `Configure regions`
* [x] Use more intuitive text for `Row width behavior` options
* [x] After creating a new flexible layout (`admin/structure/layouts/settings/flexible-template/add`) redirect user to flexible layout configuration page.
* [x] Add a `#type = help` element to the top of the flexible layout configuration form.

NICE-TO-HAVE TODOs:
* [x] Style human & machine names like we do on the content types page `admin/structure/types`
* [x] Hide the `Row width behavior` field via `#ajax`
* [x] Change order of `Row width behavior` so fluid (most common) comes first, and is default.
* [x] Put the % widths inside the rendered row examples, remove text below.
* [x] Page title for `Configure regions` should use the template name
* [x] Add `Back` buttons and `Cancel` links to multi-step dialogs
* [x] Style region width selection to remove radios entirely (like we did with layout template selector on Layout settings UI)
* [x] Make the `Number of regions` select a set of radios instead, and have them be horizontally aligned next to each other ([see mockups](https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-500186658))
* [x] Rename the `Region titles` fieldset to simply `Regions`
* [ ] Consider moving the `Selected region widths` setting into the `Regions` fieldset, as a description.
* [x] Move the `Row classes`, `Wrapper tag`m and `Row width behavior` elements into a `Style` fieldset to match the block confgiration UI. 
* [x] Move the `Style` fieldset below the `Regions` section (since it will be used less frequently)
* [x] Merge icon and name columns on layout template list
* [ ] HOLD: Style bulk operations on layout template list to match `admin/content`

NEEDS FEEDBACK TODOs:
* [ ] Add a search filter to the layout template list (does this work with VBO filters?)
* [ ] Move layout template list to new admin page `admin/structure/layouts/templates` & leave the more general "settings" page `admin/structure/layouts/settings` for general layout settings (like permissions)
* [ ] Decide on language for more intuitive text for `Row width behavior` options
    - Maximum width for each screen size (adds the 'container-fluid' class)
    - Fixed width for each screen size (adds the 'container' class)
    - Full width of page (no row classes added)
* [ ] Think about how contrib might be able to use this with other grid systems (e.g. https://getuikit.com).

FOLLOW-UP TODOs: 
* [ ] Automatically build a matching icon for each flexible layout template

---

Advocate for this issue: @klonos 

---
PR by @docwilmot: https://github.com/backdrop/backdrop/pull/2646

**Comment by BWPanda** [comment]
@docwilmot For the benefit of those not wanting/able to read through over 1,000 lines of PR code, could you please describe the feature request in more detail...? :-)

**Comment by docwilmot** [comment]
Reasonable, here goes:
Currently to create a new layout, you have to choose a template. Templates are provided in code, in a package with an info file and a tpl file, similar to modules or themes. So to create a new template requires some coding. 

This PR allows creating a new template in the UI, by adding an unlimited number of rows, with each row optionally divided into preset columns. 

The current PRs UI is probably not terribly intuitive though. To create a new template, you go to the settings page for layouts module (not the configuration page for an individual layout), click add new template, give it a unique name, then you'll be directed to an interface similar to the layout drag and drop page (it's based on this, extending the editor renderer). Once you've finished adding all the rows, save, then this new template will be available in the continue page of all layouts. Use as normal.

**Comment by herbdool** [comment]
I like where it's going.

**Comment by jenlampton** [comment]
I am so excited about this!!! Feedback follows :)

Layout settings page:
<img width="1208" alt="Screen Shot 2019-05-14 at 1 31 45 PM" src="https://user-images.githubusercontent.com/397895/57730345-23ff3d00-764d-11e9-898a-0737f8415abb.png">

* We'll need an icon for the flexible layouts (maybe in a different color?)
* There is a link `Edit` and a link `Configure`, but both of these save to config. Both should use the text `Configure`. Can we merge the two into the same form? If not, how about `Configure regions` vs `Configure name`.
* These links should be in a drop-button.

Layout builder page:
<img width="1212" alt="Screen Shot 2019-05-14 at 1 36 34 PM" src="https://user-images.githubusercontent.com/397895/57730634-be5f8080-764d-11e9-8d22-369cb0ec3784.png">

* Status message should read "Layout template created."
* Button should read "Save layout template"
* We should either change the `Top` region to `Content`, or start with the same regions as are in all the other core layout templates (header/top/content/bottom/footer).
* Layout description not used, unless I'm missing it?

Add row dialog:
<img width="632" alt="Screen Shot 2019-05-14 at 1 41 17 PM" src="https://user-images.githubusercontent.com/397895/57730798-172f1900-764e-11e9-9902-a55a628c52cb.png">

* Dialog title should be `Add Row` instead of `Add flexible layout template`
* I would prefer `Select row style` become two separate questions: `Number of columns` and `Column widths`
* Use percentages for widths, instead of # columns, for the people who don't understand that our layout is a 12 column grid.
* Change `HTML element for row wrapper` into `Row wrapper tag` to match block UI terminology
* Change `Container style` label to `Row width behavior`
* Change `Container-fluid` to `Maximum width enforced for each screen size`
* Change `Container` to `Fixed width for each screen size`
* Change `No container` to `Always full width of page`
* Change order so fluid (most common) comes first, and is default.

Row / Region naming confusion
<img width="1247" alt="Screen Shot 2019-05-14 at 1 53 16 PM" src="https://user-images.githubusercontent.com/397895/57731574-d0422300-764f-11e9-9758-e4f089d3b2e1.png">

* The row name is displaying for regions
* No way to define / change region names yet.
* What if we added region name fields to the form, and showed a # of them equal to the # of columns selected (in the same way we can show the percentage options for that number of columns). These could be optional, but if left blank then we'd show the row name instead of region name (same behavior as now).

Configure Layout:
<img width="1225" alt="Screen Shot 2019-05-14 at 1 58 51 PM" src="https://user-images.githubusercontent.com/397895/57731797-65451c00-7650-11e9-82f5-aa7eaa5860a3.png">

* Icons missing for selection area (mentioned above already)

Manage blocks: 
<img width="1184" alt="Screen Shot 2019-05-14 at 1 59 39 PM" src="https://user-images.githubusercontent.com/397895/57731841-81e15400-7650-11e9-98d2-14aa516b6ad3.png">
* Nice, here the region names are row name + number. This is a better default than on the Add row page.

**Comment by docwilmot** [comment]
Thanks for the comprehensive review @jenlampton. I agree with all your comments. A couple of replies though:
> These links should be in a drop-button.

TBH I'm not sure about the entire administrative UI for this PR, but I'm assuming you dont mind the "links within an icon" motif? But I think dropbuttons might look awful from within the icon.
> Use percentages for widths, instead of # columns, for the people who don't understand that our layout is a 12 column grid.

I considered this, but some of the fractions arent integers if you do, like `10:2` = `83.333:16.666`. Would we round to `83:17`? Is that any more intuitive than `10:2` though?
> No way to define / change region names yet.

I considered that, and it can be done, but wondered if it would be necessary, considering the region names arent exposed to the end-user.
> Layout description not used, unless I'm missing it?

😄 there is a list at `admin/structure/layouts/settings/flexible-templates`. I did say the UI needs work. I don't know where to put a link to that.

**Comment by jenlampton** [comment]
> TBH I'm not sure about the entire administrative UI for this PR

Yeah, I saw your comment about that on the main issue, but I don't have any better ideas about where to put this.

> But I think dropbuttons might look awful from within the icon.

Oh yeah, I hadn't thought of that...

> some of the fractions aren't integers ... Would we round? Is that any more intuitive than 10:2 though?

Rounding is fine, even to the nearest whole number. It is more intuitive than 10:2, but only because 10:2 only has meaning if you know it's an abbreviation of `10/12 : 2/12`. Which brings us back to the percentages :)

> I ... wondered if it would be necessary, considering the region names arent exposed to the end-user.

I noted that as a problem as well. But if we make the region names visible (and hide the row names) then people will want to change the region names. We could even kill the row names entirely, and use a random hash or something instead.

> there is a list at admin/structure/layouts/settings/flexible-templates

Ooh! I missed that. Checking it out...

**Comment by herbdool** [comment]
What if in addition to the ratio we have a visual showing the column widths? Regardless I agree it should be two dependent dropdowns: number of columns and column widths.

**Comment by jenlampton** [comment]
Maybe we could have a tiny example row?

**Comment by klonos** [comment]
Thanks for doing the first round @jenlampton ...you beat me to it 🙂  ...very thorough 👍 ...here I go now 😅 

> - We should either change the `Top` region to `Content`, or start with the same regions as are in all the other core layout templates (header/top/content/bottom/footer).

Renaming it to `Content` seems reasonable. Lets please keep things simple; 3 rows are enough to start with; people will be able add as many as they please now 🙂

> * Change `Container style` label to `Row width behavior`
> * Change `Container-fluid` to `Maximum width enforced for each screen size`
> * Change `Container` to `Fixed width for each screen size`
> * Change `No container` to `Always full width of page`
> * Change order so fluid (most common) comes first, and is default.

I say yes to this, but perhaps include the more "technical" versions of these things in brackets:

* `Maximum width enforced for each screen size (container-fluid)`
* `Fixed width for each screen size (container)`
* `Always full width of page (no container)`
* `Row width behavior (container style)`

>>> - These links should be in a drop-button.
>>
>>TBH I'm not sure about the entire administrative UI for this PR, but I'm assuming you dont mind the "links within an icon" motif? But I think dropbuttons might look awful from within the icon.
>
>Yeah, I saw your comment about that on the main issue, but I don't have any better ideas about where to put this.

I think that it's high time we switched from the tile listing to a table here. This would allow us to place the links in dropdown menus, within an "Operations" column (which is a common UI pattern we use elsewhere). It would also allow for a search/filter section at the top, which people could use to search the list of templates; and we could have filters like a select for 1/2/3 column layouts etc (I think that the D8 Layout Builder allows that too ...or maybe I am thinking [Panelizer](https://www.drupal.org/project/panelizer)?).

...we can leave the template selection when editing/creating layouts still be a tile listing; that's fine.

>> Layout description not used, unless I'm missing it?
>
> 😄 there is a list at `admin/structure/layouts/settings/flexible-templates`. I did say the UI needs work. I don't know where to put a link to that.

See my comment above. Having a table would allow showing the description there (as well as allow for a column with a list of which layout each template is being used in).

> - There is a link `Edit` and a link `Configure`, but both of these save to config. Both should use the text `Configure`. Can we merge the two into the same form? If not, how about `Configure regions` vs `Configure name`.

Yes, lets please just merge the edit and configure forms into one, with the name/description fields being at the top of the region management UI (a-la #475). No need to have people jump into another for to simply rename a layout template.

Some other suggestions:

- Move the "Only templates that are selected here will be available when configuring layouts." message at the top of the page in `/admin/structure/layouts/settings`. As it is now, as more layouts get added, this text gets pushed down, outside the viewport.

- We shouldn't call them "flexible"; I'd prefer "custom" instead. Later on, perhaps we can allow core layout templates to be edited too; they would have an "overridden" state, and a "revert" action.

- "Deleting the template will affect any layouts using this template." this confirmation dialog could use a list of layouts using it. If not used anywhere, is there any point in having a confirmation dialog? ...we could say "This tamplate was not used in any layout" in the confirmation message.

- (follow-up to the previous point) What happens to layouts that were using a custom template when that template gets deleted? Do we switch them to another template? ...if so, perhaps have the user choose another template as part of the deletion workflow? Either that, or disallow deleting custom templates while they are being used.

- Canceling a custom layout template deletion redirects to `/admin/structure/layouts/settings` instead of `/admin/structure/layouts/settings/flexible-templates` ...which is not a bad thing at all, if we decide to turn that into a table and include the operations drop-downs there.

- When clicking the "Add row above" action, the drop-down is hidden behind the row:

<img width="587" alt="Untitled" src="https://user-images.githubusercontent.com/2423362/57741451-3ab29c00-7700-11e9-922a-bbb3b4f54d75.png">

- Nothing happens visually when you remove a row (either existing, or newly-added). You see the effect of the deletion only after saving the layout template.

- The name of the layout template used for each layout in `/admin/structure/layouts` is a link. It currently leads to the configure layout form for each layout. Now I think that perhaps it makes sense to make it point to the edit page of that template.

That's it for now. Loving this ❤️ ...so much so, that I am tempted to set the milestone to 1.14 😄

**Comment by docwilmot** [comment]
> Lets please keep things simple; 3 rows are enough to start with; people will be able add as many as they please now

Agreed, plus, once you've added new rows, they wont be necessarily at the "top" or "bottom" any more!
>Maximum width enforced for each screen size (container-fluid)
Fixed width for each screen size (container)
Always full width of page (no container)
Row width behavior (container style)

I think these are too verbose, and still unclear; how about:
- Fixed width (container)
- Full width with margin (container-fluid)
- Full width without margin (no container)
- Row width behavior (container style)

> I think that it's high time we switched from the tile listing to a table here

That would work nicely, especially if it mimics the layout list in appearance, but it would be a big change, would like to hear thoughts on that!
> Yes, lets please just merge the edit and configure forms into one

I dont like this though, this UI borrows from the Editor renderer which builds the drag and drop block page for layouts: an initial page to create the layout and a separate page to configure it. And yes it matches other areas in core where you define a thing with its machine name, then you go somewhere else to configure it (menus, Views, etc). Plus, did I mention I dont like it? I think it would look awkward. 
> We shouldn't call them "flexible"; I'd prefer "custom" instead

I considered that, and I also dont really like "flexible" but "custom" is taken: thats what you'd call templates you make and store in `/layouts`, like custom themes or modules.
> What happens to layouts that were using a custom template when that template gets deleted?

Good point; could swap them to whatever "Default layout" is using, or Moscone.
> Nothing happens visually when you remove a row (either existing, or newly-added). You see the effect of the deletion only after saving the layout template.

Missed that. Fixable.

**Comment by olafgrabienski** [comment]
Wow, I also like the PR!

(My test page: http://2646.backdrop.backdrop.qa.backdropcms.org/test-page)

I've only one question at the moment: While adding rows to a test layout, I was wondering if I can reorder them, and it doesn't seem to be possible. Is that by design?

**Comment by jenlampton** [comment]
> Lets please keep things simple; 3 rows are enough to start with ... Agreed, plus, once you've added new rows, they wont be necessarily at the "top" or "bottom" any more!

👍 

> I think these are too verbose, and still unclear; how about:
> Full width with margin (container-fluid)
> Full width without margin (no container)

These options are less clear; a margin can be anything! I'd much rather have an accurate but verbose description that people will understand, than something that's too short and ends up cryptic. 

* Maximum width for each screen size (adds the 'container-fluid' class)
* Fixed width for each screen size (adds the 'container' class)
* Full width of page (no row classes added)

> it's high time we switched from the tile listing to a table ... This would allow us to place the links in dropdown menus, within an "Operations" column

But [we already have a table](http://2646.backdrop.backdrop.qa.backdropcms.org/admin/structure/layouts/settings/flexible-templates). and making "Which templates are allowed?" question into a giant table makes it hard to tell that it's even a question anymore. 

> we can leave the template selection when editing/creating layouts still be a tile listing

We can leave the template selection for enabling/disabling layout-templates too. 

The tiles on the settings page are intended for one purpose: to enable/disable the layouts allowed, and to give people a visual cue (the tile) as to which is which. I think we should leave this interface as-is, and move all the other layout-template related things to the layout-template listing page, the one that's a table. 

(That includes moving the action link `Add flexible layout template` -- or whatever we decide to call it -- to that page too)

> lets please just merge the edit and configure forms into one

I have the same hesitation about this as I do with #475. It would only be an improvement with a simple layout. The same way that is only an improvement when you menu is less than 50 links long. When you have a complex layout, or a very long menu, you don't want those other settings on the same page, or in the same form.

> Move the "Only templates that are selected here will be available when configuring layouts." message at the top of the page in /admin/structure/layouts/settings

This change isn't necessary if we use the existing layout-template table UI for everything this form element wasn't intended to do.

> We shouldn't call them "flexible"; I'd prefer "custom" instead. ... I considered that, and I also dont really like "flexible" but "custom" is taken: thats what you'd call templates you make and store in /layouts, like custom themes or modules.

I agree with @docwilmot, having another namespace conflict for layouts would be a bad idea. [For WordPress](https://www.webdesignerdepot.com/2015/10/edit-any-wordpress-theme-with-divi-builder/) these are called `Divi Layouts`. Can we steal that name?

> Now I think that perhaps it makes sense to make it point to the edit page of that template.

I disagree. I don't think we should take a tool people might have already been using and change it to do something else. It would be better to add another link. 

> What happens to layouts that were using a custom template when that template gets deleted?

Can we check and see if a layout-template is in use, and throw an error instead? I expect that this would be a mistake or an oversight 99% of the time. I'd prefer that we make people actively change their layout-template before they are allowed delete one that's in use.

> I was wondering if I can reorder them, and it doesn't seem to be possible. Is that by design?

I expect this can be added, but it's probably not easy!

---

There's a bunch of great feedback in this issue already, but I want to stop us from having too many more good ideas until we get a round 2 of the PR. Otherwise, there will be too much to keep track of in these comments. (Maybe we can create a todo-list of changes in the parent issue?)

**Comment by klonos** [comment]
> But we already have a table. and making "Which templates are allowed?" question into a giant table makes it hard to tell that it's even a question anymore.

No it won't; we'll gray the disabled ones out, and move them to the bottom of the table; same as we do with the views listing for example 😉

> These options are less clear; a margin can be anything! I'd much rather have an accurate but verbose description that people will understand, than something that's too short and ends up cryptic.
> 
> - Maximum width for each screen size (adds the 'container-fluid' class)
> - Fixed width for each screen size (adds the 'container' class)
> - Full width of page (no row classes added)

> There's a bunch of great feedback in this issue already, but I want to stop us from having too many more good ideas until we get a round 2 of the PR. Otherwise, there will be too much to keep track of in these comments. (Maybe we can create a todo-list of changes in the parent issue?)

Yes, I will go through all comments and make a tasks/ideas list in the issue summary. For the ones that we don't have consensus, I'll add a "Decide if ..." prefix 😉

**Comment by jenlampton** [comment]
> I will go through all comments and make a tasks/ideas list in the issue summary

Perfect, thank you! 

> For the ones that we don't have consensus, I'll add a "Decide if ..." prefix 😉

"needs more feedback" maybe? :)

> we'll gray the disabled ones out, and move them to the bottom of the table; same as we do with the views listing for example

I think my resistance to this idea is that right now we only have one layout setting: "Which layout templates?" ...but the current page is a layout settings page (and not a layout template list) because we were anticipating having more settings for layouts. Especially as we start tp expand layouts to match more of what panels provided. 

Where do the other setting questions go, if we remove the layout settings form and replace it with a layout template list? Or do we remove the form now, but add it back later, when we have more layout settings?

**Comment by docwilmot** [comment]
The layout templates settings page `admin/structure/layouts/settings`, with a table instead of icons. 
Still WIP, and havent sorted out default checked checkboxes for enabled templates:

![Layout settings   PR 2646 backdrop backdrop testing site](https://user-images.githubusercontent.com/1042283/58708079-5e224b80-837c-11e9-99d5-42fbb87feb4b.png)

**Comment by klonos** [comment]
Looking good @docwilmot 👍 ...as always ❤️ you work!

Thoughts/ideas following...

**Comment by klonos** [comment]
![Screen Shot 2019-06-01 at 12 25 45 am](https://user-images.githubusercontent.com/2423362/58712420-d5e78900-8403-11e9-84a9-e2989ab59ac9.png)

- the page title could use the template name, so I know what I'm editing.

- we could get rid of the 2 "add row above/below" actions from each dropbutton, and instead add "+ Add row" buttons; something like this perhaps:

![Screen Shot 2019-06-01 at 1 03 10 am](https://user-images.githubusercontent.com/2423362/58714913-1e557580-8409-11e9-9b55-8e5a8b77c37a.png)

**Comment by klonos** [comment]
...back to the layout settings page (layout template table):

- we have removed the container/border from the bulk operations in the rest of the UI, we should do the same here

- we could use a search filter

- "Configure" should be the default operation for user-created templates

- I don't see the point in having the layout icon be a separate column. Merge with "Name"

- Make human names bold, and add machine names in a separate row. Like we do in the Views listing page

**Comment by docwilmot** [comment]
Thanks for the feedback. Not sure I like the 2nd screenshot in https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-497743027, will see what everyone thinks, but it seems a bit more complicated than I'd like.

**Comment by docwilmot** [comment]
But I've just realised a serious weakness of my current UI; you cant edit column names in a row! 😮

**Comment by olafgrabienski** [comment]
> we could get rid of the 2 "add row above/below" actions

I'd also prefer an alternative to the "Add row above / below" actions. The two options were confusing me slightly, it didn't seem clear to me on first sight, where "above" and "below" exactly would be. Anyway, in each case, it would be helpful to have the option to reorder the rows.

**Comment by docwilmot** [comment]
updated the PR;
- added a hover to each row, so that the "Above" becomes obvious. 
- changed the order of adding a new row: now you pick the style first, then do the name and other configuring. More intuitive, and much less code too.

haven't addressed https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-497745486 yet
and will need to think a UI for reordering rows.
let me know what y'all think so far.

**Comment by olafgrabienski** [comment]
@docwilmot I like the direction! Unfortunately I don't have much time over the weekend, therefore only one quick response:

If we specify an "Above" or "Below" to the "Add row" link, I'd prefer "Add row below". AFAIK putting things *below* is the usual pattern in Backdrop, when you add something, e.g. fields or filters in Views, or blocks on a Layout.

The latter, blocks on a Layout, could also in other aspects be a model for the handling of rows on a Flexible layout page, so we'd have more consistency. That would mean: no "below" or "above" in the "Add" link, but just putting the new row below, which is fine, if you're able to reorder them.

**Comment by docwilmot** [comment]
@olafgrabienski I just understood what you meant. That's a good idea maybe. Just an "Add"button, then drag and drop to rearrange. Worth considering.

**Comment by klonos** [comment]
Comments/thoughts on [the recent changes](https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-499903826):

- I like the hover indication for the rows, but the "animation" annoys me. perhaps the dotted "container" can be there permanently, and the background color is all that changes (without anything "moving"). Also, we are using `#ffff99` for indicating hovers elsewhere in core; should we be using the same here?

    ...having said that, I still think that having "add row here" links would work better, because a) the user does not need to think much which row to select the action from, b) less clicks, and c) allows the "configure row" action to become the default in the dropbutton. In my head, the "add row" action should be a layout template action; having it in the same dropbutton as the configure/delete row actions "ties" it to the specific row.

- I absolutely love the implementation of selecting row style first 👍 ...especially the visual representation of the styles.

**Comment by klonos** [comment]
...a few quirks to iron out:

- This is what the "add row" dialog looks like:

  <img width="623" alt="Screen Shot 2019-06-09 at 9 10 16 am" src="https://user-images.githubusercontent.com/2423362/59153163-8f92c980-8a96-11e9-8101-97e0c7f71107.png">

- There is no point in having the "select number of columns" option in the dropdown; and I would expect the "Column style" field to be hidden via `#states` when no selection is made. Lets remove that option instead, change the label of the drop-down to "Number of columns", and have 1 be the default.

- If "1 column" is the selected option, then also select the only checkbox, so that the user does not have to click 2 times. In fact, lets preselect the first radio of all options. This would help avoid this (or having to add validation for it):

  <img width="624" alt="Screen Shot 2019-06-09 at 9 38 26 am" src="https://user-images.githubusercontent.com/2423362/59153345-72f89080-8a9a-11e9-8e88-bdf9d064b005.png">

- Place the column style labels next to the radios, to save vertical space on the dialog (less scrolling):

  <img width="622" alt="Screen Shot 2019-06-09 at 9 20 30 am" src="https://user-images.githubusercontent.com/2423362/59153245-52c7d200-8a98-11e9-9e2f-1f113d0aff2d.png">

- Make it so that the entire machine name `<span>` is wrapped under the row name field.

- Add a "back" link/button to the "Add row" dialog, to return to selecting number of columns and/or style.

- Add a "cancel" link to both "Add row" and "Configure row" dialogs.

- Consider moving the "Selected column style" field closer to the "Column titles" fieldset, or even making it a description for it. Alternatively, we can included it in the dialog title.

**Comment by klonos** [comment]
I have added myself as an advocate for this issue for 1.14. I will be providing feedback, requesting additional feedback from others, and providing a status summary during our weekly dev meetings.

**Comment by klonos** [comment]
...also add row names in `/admin/structure/layouts/settings/[layout_template]/configure` page.

**Comment by docwilmot** [comment]
> I still think that having "add row here" links would work better

I'm resisting this because I feel it complicates the UI. This would require *two* dropbuttons for each row, (or dropbutton and a link as in your earlier comment). I like the simplicity of the current design, but I'm not designing for Apple, so I'm open to consensus.

p.s. its still only one click to add a new row, just use the "Add above" link below where you want your new row.

> ...also add row names in /admin/structure/layouts/settings/[layout_template]/configure page.

Thats already done, unless I'm misunderstanding.

Everything else, thanks, will do.

**Comment by klonos** [comment]
> ...but I'm not designing for Apple, so I'm open to consensus.

😆 

> p.s. its still only one click to add a new row ...

Not if you want to add a row after the last one. And it makes "Configure", which is the main action for rows be 2 clicks away.

> Thats already done, unless I'm misunderstanding.

When no column titles have been specified, the region names fall back to the name of the row. So in the PR sandbox where I tested things, I edited the `test_again` layout template, and I added titles to all columns. initially for the 1-column rows:

<img width="942" alt="Screen Shot 2019-06-09 at 2 06 42 pm" src="https://user-images.githubusercontent.com/2423362/59154928-f1682900-8ac0-11e9-80f7-ea09608b7e0b.png">

This became more "interesting" when I removed the titles of all 3 columns in the "Test row" row:

<img width="1174" alt="Screen Shot 2019-06-09 at 2 07 52 pm" src="https://user-images.githubusercontent.com/2423362/59154958-5faceb80-8ac1-11e9-822d-0c46d3828ed9.png">

...I would expect the column names to be "Column 0"/"Column 1"/"Column 2", but they got "test_row 0"/"test_row 1"/"test_row 2" as their titles instead.

PS: should we be calling the columns "regions" by the way?

**Comment by klonos** [comment]
Mockups for the 1st dialog, where the user is asked to select region styles:

<img width="626" alt="Screen Shot 2019-06-09 at 3 25 49 pm" src="https://user-images.githubusercontent.com/2423362/59155395-e4046c00-8acb-11e9-8ec6-d4fd2a698809.png">

<img width="620" alt="Screen Shot 2019-06-09 at 3 39 41 pm" src="https://user-images.githubusercontent.com/2423362/59155425-d7344800-8acc-11e9-88ee-c315465715e3.png">

- Dialog title becomes "Add regions"
- Call columns "regions" instead (because that's what they are)
- Make the "Choose column style" select a radio set instead, and call it "Number of regions"
- Call "Column style" "Region widths" instead
- Hide the radio input for selecting width; make the width images work the same way as tamplate tiles work currently. Move the labels above each tile and center them horizontally
- Selected tile and hovered-over tiles get a `#43afe4` border.

**Comment by mazzech** [comment]
Wow, cool to see this moving forward... I am happy to help testing once it's in Alpha:-). Just one question: will this be framework-agnostic or based on Bootstrap CSS? Because my preferred framework handles grids quite differently:

https://getuikit.com/docs/grid#child-width

However, it's still possible to overwrite the blocks per screen size, which I use quite a lot. 
https://getuikit.com/docs/width#responsive-width

Maybe we can take this approach into account... on the other side, I also think that ease of use must be a top priority. Or perhaps additional CSS frameworks can by supported by module

**Comment by jenlampton** [comment]
@mazzech core will always only support Bootstrap, but you raise a good point that we should test how contrib might be able to plug in to the flexible layout. It might be a bit early for that, but I'll add a todo to the parent issue so we don't forget!

**Comment by jenlampton** [comment]
I red through the entire list of comments here, and it looks like @klonos recommended some of the same changes I did in my first review, almost 30 days later, and @docwilmot came to some of the same realizations i posted in my first review, also days later. It's obvious that we've all been having the same good ideas -- but also that we can't read this many looong comments :)

So ...  

I've updated the top issue withe a few to-do lists. 

* ACTONABLE TODOs are things we all agree on (or at least, no objections have been stated as of the last recommendation -- we can move these to the NEEDS FEEDBACK list if not)
* NICE-TO-HAVE TODOs are things that we don't need to get done before merging into core
* NEEDS FEEDBACK TODOs are things we should still discuss.

**Comment by docwilmot** [comment]
flippin heck @jenlampton what a checklist, well done.

**Comment by docwilmot** [comment]
>  Move layout template list to new admin page & Restore settings page for general layout settings (like permissions)

I dont understand what this means @jenlampton

**Comment by mazzech** [comment]
@jenlampton great action list:-) Regarding non-Bootstrap themes, maybe it's even sufficient to have a field for additional CSS, as I have seen on a mockup earlier.

**Comment by klonos** [comment]
> * [ ] Remove text `Select number of columns` from select list (we should check elsewhere in core, but I think the top item is usually just `- Select -`)

I wonder how people feel about changing this drop-down menu to a series of horizontally-aligned radios instead, as per my previous mockups:

<img width="210" alt="Screen Shot 2019-06-12 at 5 28 18 am" src="https://user-images.githubusercontent.com/2423362/59300616-f4d2fe80-8cd2-11e9-8a85-372a48d6aa30.png">

...having a drop down seems like a waste of space/clicks when the options a 1-character long (1/2/3/4).

**Comment by jenlampton** [comment]
Like the suggestion @klonos, can you add that to the nice-to-have list?

**Comment by jenlampton** [comment]
> Move layout template list to new admin page `admin/structure/layouts/templates` & Restore settings page `admin/structure/layouts/settings` for general layout settings (like permissions)

@docwilmot I guess I was just recommending a different URL for the layout template page... something like `admin/structure/layouts/templates` as opposed to `admin/structure/layouts/settings`. The more general `settings` path could be reserved for more general settings (like the permission matrix).

**Comment by docwilmot** [comment]
>  When no region name is specified, display a label with an auto-increment number, like "Region 1"/"Region 2"/"Region 3" (start from 1 and not zero). Numbers should not reset per row.

a thought: if you have 4 regions, numbered as recommended here, and you give only the first one a name (header, for example) which of the following would apply:
- header, region 2, region 3, region 4 OR
- header, region 1, region 2, region 3

the first would ensure remaining regions keep their names if you change or rename one.

**Comment by docwilmot** [comment]
> @docwilmot I guess I was just recommending a different URL for the layout template page... something like admin/structure/layouts/templates as opposed to admin/structure/layouts/settings. The more general settings path could be reserved for more general settings (like the permission matrix).

this makes near perfect sense, except that we have no settings, this page has always been just which templates are enabled. i think changing the page to a table makes it more settings-y though.

**Comment by jenlampton** [comment]
But it makes it harder to add other settings, which we do eventually want
to do. Permissions is the one thing we'll want to do right away, but I
expect there will be other things, maybe related to contexts,
relationships, or who knows what contrib might add.

**Comment by klonos** [comment]
> ...add that to the nice-to-have list?

Done 👍

**Comment by docwilmot** [comment]
@klonos @jenlampton see https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-501114386. That'd be a big change in code if we wanted this, so let me know what you think.

**Comment by klonos** [comment]
@docwilmot I think that this can be a follow-up. Lets keep things simple; baby-steps to get an MVP ASAP - then iterate to improve. So for this feature specifically, lets simply make it so that region name fields have a default value of `Region x` (where x is an auto-incremented number starting from 1), instead of being blank.

**Comment by docwilmot** [comment]
Maybe you misunderstand: currently region names do autoincrement, but only within rows. This proposal from @jenlampton recommends they autoincrement ignoring rows. My point in https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-501114386 is that this would result in confusion, as I illustrated, if persons add rows. Unless there is an idea on how to mitigate this, I'm proposing to autoincrement only within rows.

**Comment by jenlampton** [comment]
> a thought: if you have 4 regions, numbered as recommended here, and you give only the first one a name (header, for example) which of the following would apply

The provided name would appear in the UI instead of the auto-generated name: header, region 2, region 3, region 4. Region 1 is now named "Header". (The machine name can remain `region_1`).

> what happens if they add a new row in between, or change the number of columns in a row?

Then new region numbers appear. If they don't like the out-of-order problem, they can name the regions. (I expect most people will name their regions anyway - we just need a sensible default for those that don't)

> My point in #3755 (comment) is that this would result in confusion

It would be even more confusing if they auto-incremented only within rows, as a layout could end up with several different regions all named "Region 1". All the region names within a layout need to be unique so people can move blocks from one region to another. 

> We cant make the rows re-order because there might be all kinds of CSS linked to region identifiers and tags and whatnots.

As long as the region machine names are unique and never change (region_1, region_2, etc) the class names will also be unique and never change (region-1, region-2). There will only be problems with CSS if people choose to add their own custom classes, and attach their own custom CSS -- neither of which we need to worry about in core.

**Comment by docwilmot** [comment]
I'm still in disagreement that this is better.

I'll clarify the issue for those who come across this here: We have two options for if you create a flexible layout template **and you dont explicitly name all the regions in it.**

Consider you start with a simple layout template with a two colum first row, then a three column second row.

The first option, the current status of this PR, is that the regions will be named according to the row theyre in
```
Header 1 | Header 2 
Content 1 | Content 2 | Content 3
```

The second option is to number all regions consecutively
```
Region 1 | Region 2
Region 3 | Region 4 | Region 5
```

If you add new rows in between, you'll end up with this
```
Header 1 | Header 2 
AnotherRow 1 | AnotherRow 2 | AnotherRow 3
NewRow 1 | NewRow 2
Content 1 | Content 2 | Content 3
```
vs
```
Region 1 | Region 2
Region 8 | Region 9 | Region 10
Region 6 | Region 7
Region 3 | Region 4 | Region 5
```

In the second scenario of course we can force sequential renaming to end up with 
```
Region 1 | Region 2
Region 3 | Region 4 | Region 5
Region 6 | Region 7
Region 8 | Region 9 | Region 10
```
But then if a user had some blocks in the region originally named "Region 3" that region would now be renamed "Region 9". If that user had had a CSS declaration `.region-3 .important-block`, that breaks immediately.

I can understand that it would be nice to have regions sequentially named, but I for one see no evil in having regions named according to their rows, and I don't think its worth the dissonance I feel looking at the craziness that could ensue when adding rows.

Furthermore, we could argue that users should be forced to explicitly name all their regions, but this would be at the expense of the ease of having Backdrop automatically name them for you.

So I prefer the first method/option, for better UX and more sanity.

**Comment by jenlampton** [comment]
@docwilmot you are starting with the assumption that rows will have names. There is absolutely no reason for a row to have a name since it will never be visible to the end-user. So the example of:
```
Header 1 | Header 2 
Content 1 | Content 2 | Content 3
```
will actually be:
```
EqquJNc 1 | EqquJNc 2 
oqZ7pDp 1 | oqZ7pDp 2 | oqZ7pDp 3
```
So people should comapre that, to something with numbred regions, like:
```
Region 1 | Region 2 
Region 3 | Region 5 | Region 4
```

**Comment by jenlampton** [comment]
Here's a thought: If the only problem is that the Region names have a sequence, maybe we should choose something that's not a number for the region names? Can we think of anything else that has an infinite (or at least very large) set? Something like colors:
`Region red`
`Region red`
`Region red`

**Comment by docwilmot** [comment]
Currently rows have names. Are you suggesting we get rid of that?

![rowsnames](https://user-images.githubusercontent.com/1042283/59398269-2c828a80-8d55-11e9-8d0c-f07a09b42be7.PNG)

**Comment by jenlampton** [comment]
Yep. In Backdrop, row's don't have names. This field asks people to answer a question that doesn't need to be asked.

**Comment by klonos** [comment]
I am in agreement with @jenlampton: rows should only have machine names, which should be auto-generated (use UUID even?). I see no reason to be exposing that to the end users at all.

As for region names, we should make them required fields, but with automatically-generated default values for their human names. If users don't like those, they'll have the chance to change them. This solves the problem of not providing any names, while at the same time permits the user to skip that step (they will be able to rename regions later if they wish to).

**Comment by jenlampton** [comment]
> As for region names, we should make them required fields...

@klonos did you mean we should **not** make them required fields?

**Comment by klonos** [comment]
No I meant required, and pre-filled with auto-generated values. The user can skip changing those defaults, or do change them to whatever they like; but if they blank the fields out, there should be a validation error. There should not be regions without machine and human names; right?

**Comment by docwilmot** [comment]
New PR up, 
- Removes row titles. 
- Adds an auto-generated title for regions as a placeholder (it could also be as default value but placeholder saves having to delete the default value first). Temporarily using `uniqid()` to generate a random string, but I'm thinking we could generate user-readable strings instead, maybe re-use this code here https://github.com/backdrop/backdrop/pull/1763/files#diff-0d9bd97d91568a099e773bc296444be2R971.

**Comment by docwilmot** [comment]
>  Default drop-button action for flexible layout should be Configure regions

Is this on the current layout `admin/structure/layouts/settings` page we're talking about? (if it is, the default action for that page has been to enable or disable layout templates. Should we not maintain that for all templates? Not sure)

>  A single 'Add new row' button at the bottom

Does this imply completely get rid of the ability to "Add a row above/below", only add new rows to the bottom (or top), and depend on drag/drop to re-order?

**Comment by klonos** [comment]
> Is this on the current layout `admin/structure/layouts/settings` page we're talking about?

I think yes.

> Does this imply completely get rid of the ability to "Add a row above/below", only add new rows to the bottom (or top), and depend on drag/drop to re-order?

I also think yes. This basically:

![image](https://user-images.githubusercontent.com/2423362/59553339-0c61fe00-8fd7-11e9-8961-bc4b1b9d9dda.png)

**Comment by docwilmot** [comment]
> I also think yes. This basically:

I'm interpreting this differently, more like @olafgrabienski's idea to have no links to add rows at all, neither on the dropbuttons, nor in-between like your mockup, and just have *one* "Add row" at the bottom, then drag the row where you want it.

Need that to be clear, since it will require a bit of re-writing.

**Comment by olafgrabienski** [comment]
> I'm interpreting this differently, more like @olafgrabienski's idea to have no links to add rows at all, neither on the dropbuttons, nor in-between like your mockup, and just have one "Add row" at the bottom, then drag the row where you want it.

@klonos For reference, [my idea](https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-500167914) was to treat adding rows the same as adding blocks on a Layout: no "below" or "above" (and no "here") in the "Add" link, just put the new row below (and allow to reorder rows).

@docwilmot With adding blocks to a layout as pattern, the "Add row" link (and other links) would be in a dropbutton beside the region name, see the screenshot of the *Manage blocks* page:

<img width="310" alt="screen-backdrop-add-block-to-layout" src="https://user-images.githubusercontent.com/14188307/59554451-487b6b80-8fa3-11e9-9c4e-5484d84d1590.png">

**Comment by jenlampton** [comment]
@olafgrabienski I'm not sure the blocks model there works for rows... Are you recommending that we add a new "Add row" link into every region? Like this?
<img width="1254" alt="too-many-add-rows" src="https://user-images.githubusercontent.com/397895/59555386-11d33000-8f66-11e9-961f-637a31597d18.png">

I was thinking we could so something more similar to fields:
<img width="1195" alt="Screen Shot 2019-06-15 at 12 01 42 PM" src="https://user-images.githubusercontent.com/397895/59555347-878acc00-8f65-11e9-8826-e1dca52223a8.png">

**Comment by docwilmot** [comment]
OK we all have different ideas it seems. I was more thinking that @olafgrabienski meant this (and I dont mind that as an idea either):

![add row button](https://user-images.githubusercontent.com/1042283/59555464-3afcbc00-8f78-11e9-9070-9ed25c9cc35f.PNG)

**Comment by jenlampton** [comment]
I like that too :) @klonos are you okay with this as a first iteration?

**Comment by olafgrabienski** [comment]
@jenlampton Good point, I just didn't reflect that there can be several regions in a row!

@docwilmot I like the big "Add row" button at the bottom.

**Comment by klonos** [comment]
Yes, I am fine with the single "Add row" thing as a first iteration ...but not as a big submit button at the bottom. We currently have a UI pattern for adding things, and that is the "+ Add thingy" link at the top-left of the page.

@jenlampton we even have plans to change the "add new filed" draggable thingy thing into a "+ Add field" link: #779 ...so I would like to see us use that pattern consistently throughout the UI.

**Comment by docwilmot** [comment]
You say "as a first iteration"as if you had an ultimate better standard. Since this is new, best get it done the first time. What would be the best last iteration?

And the + Add thingy links don't seem right for this.

**Comment by klonos** [comment]
...having said the above, I would like to point the following:

- if too many rows are added in a flexi layout, then having a single button to add new rows (be it at the top or bottom of the page), becomes a UX issue in smaller screens. The user will need to scroll up or down to find that button, and then drag things around. Having "Add row here" links solves those issues.

- in the "Manage blocks" page for each layout, the "Add block" action is the main, most common one for regions (the only thing that people can do with the "Configure region" action, is to tweak wrapper tags and add classes). So we don't have any need to move that out of the drop-button there, but we do here.

- yes, one can get into that scrolling/dragging UX situation in the "Manage blocks" page, but this happens only in narrow regions (sidebars), with many blocks. I don't find that to be very common.

- if the "Add row here" links are too much, the I would like us to reuse an  existing UI pattern, like the +/- buttons in the options element:

![Screen Shot 2019-06-16 at 8 46 44 am](https://user-images.githubusercontent.com/2423362/59557079-a6479c00-9013-11e9-89b4-0d44ee7fb0aa.png)

...something like this perhaps (quick mockups - think pretty +/- buttons instead):

![Screen Shot 2019-06-16 at 9 12 48 am](https://user-images.githubusercontent.com/2423362/59557243-312a9580-9018-11e9-8a11-8a2567a7567c.png)
![Uploading Screen Shot 2019-06-16 at 9.21.02 am.png…]()

**Comment by klonos** [comment]
> And the + Add thingy links don't seem right for this.

Care to elaborate @docwilmot? is it because the "Add row" is popping up a dialog instead of loading a separate page? If so, then you are right, and I'll have to think about the UI/UX a bit more.

...but having it as a submit button is even more confusing IMO.

**Comment by klonos** [comment]
> You say "as a first iteration"as if you had an ultimate better standard.

I don't 😅 ...I currently only see the UX issues. Just trying to brainstorm solutions with you guys. Sorry if I am being think/annoying 😅

**Comment by docwilmot** [comment]
No no, it's me. Sometimes I speak too directly, the wife says. I meant that as a literal question, not a complaint. You are in now way being annoying. I really would like to hear what ideas you have, anyone has. We haven't all yet agreed so now's the time.

**Comment by jenlampton** [comment]
> You say "as a first iteration"as if you had an ultimate better standard. Since this is new, best get it done the first time. 

I also strongly disagree with this. If we wait to get it right the first time, we'll never get anything done. **better is the enemy of done**. We need to get it done **first**, then we iterate and make it better. Besides, we aren't ever truly gong to know what is best until we actually use it. :)

@docwilmot I'd lean towards doing whatever is easiest for you to do initially (even if that is nothing) to get the rows draggable, just so we un-block the ability to get flexible layouts into core. 

That said, the [example I like most](https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-502392557) is @docwilmot's visual of what @olafgrabienski might have been recommending. It's in line with what I was thinking, and it's also simple.

I dislike [all the recent suggestions](https://github.com/backdrop/backdrop-issues/issues/3755#issuecomment-502406522), especially the ones with links on the right. It makes the UI feel cluttered and overwhelming.

**Comment by jenlampton** [comment]
@docwilmot I've unchecked some of the previously-checked items on the master (top) list because they weren't working for me in the sandbox, but it occurs to me that you might have a more recent iteration that has those changes... If so I promise to re-check all the boxes when I can test it :)

**Comment by klonos** [comment]
> I dislike all the recent suggestions, especially the ones with links on the right. It makes the UI feel cluttered and overwhelming.

Awww 😞 ... 😄 ...but I said "quick mockups - think pretty +/- buttons instead". Perhaps I should actually finish my mockups if I want to "sell" my ideas 🙂

**Comment by klonos** [comment]
...if we are going with the single submit button at the bottom for adding rows, can we at least move it to the right (left in RTL), so that it has a distance from the other submit buttons in the form?

**Comment by jenlampton** [comment]
The + - buttons will mean nothing without context. I'd give up pretty any
day for something more usable.

I don't mind the add button on the left, it's a different color and won't
be confused with submit. On the right it will get overlooked.

**Comment by klonos** [comment]
I guess that we do have a pattern where such a button is to the left:

![Screen Shot 2019-06-16 at 11 29 32 am](https://user-images.githubusercontent.com/2423362/59557959-1a8d3a00-902a-11e9-957e-193d31de25f6.png)

...the only difference there is that there is something in between to separate that button from the main submit buttons in the from.

**Comment by docwilmot** [comment]
I've got most of the important stuff done. Sandbox will surely need rebuiding after all tthese changes, but rushing off to work now.

**Comment by klonos** [comment]
Absolutely loving this!! ❤️:

<img width="614" alt="Screen Shot 2019-06-20 at 5 22 59 am" src="https://user-images.githubusercontent.com/2423362/59794168-b1544200-931b-11e9-879b-f465cb07a06a.png">

**Comment by olafgrabienski** [comment]
Wow, and reordering rows works now! Didn't test thoroughly yet, but looks goood at first sight.

**Comment by klonos** [comment]
>Wow, and reordering rows works now!

...aha! I did not realize that. Perhaps it'd be a better indication that this is the case, if there were crosshair drag handlers on the left of each row (right on RTL)? As implemented currently, the cursor changes to a crosshair when you hover over regions (which is why I missed it) - not when hovering over the row. This may give the false impression that regions can be moved around.

**Comment by klonos** [comment]
...perhaps we should be using some `text-overflow: ellipsis;` on the regions:

<img width="1193" alt="Screen Shot 2019-06-20 at 5 35 24 am" src="https://user-images.githubusercontent.com/2423362/59798873-73105000-9326-11e9-984e-8d1c57978929.png">

**Comment by docwilmot** [comment]
At this stage, I need to bikeshed all remaining issues, unfortunately.
>  Add a unique icon to use for flexible layouts (can we change the color, maybe?)

Will need a designer person to handle this
> Default drop-button action for flexible layout should be Configure regions

Not sure I agree; the default for other templates will be enable/disable, so why be different?
> Use more intuitive text for Row width behavior options

Already discussed, not settled
> Style bulk operations on layout template list to match admin/content

Not sure what this is asking for sorry
>  Style human & machine names like we do on the content types page admin/structure/types

Not sure what this is asking for sorry
>  Merge icon and name columns on layout template list

Doable, but necessary? (least difficult point)
>  Hide the Column style/widths field via #ajax

Not sure what this is asking for sorry
> Consider moving the Number of columns setting into the fieldset, as a description.

Not sure what this is asking for sorry
> Add Back buttons and Cancel links to multi-step dialogs

Need some advice here. 
- When adding a *new* row, we start with region widths dialog, so *there will need to be a cancel button on the region widths dialog*
- then we move to the row config dialog. Should there be a cancel there, or a back? If cancel, should it reload the region widths dialog or just close the dialog?
- When configuring an existing row, we start with the row config dialog. A cancel button should be there now, I'd say
- then if you want to change the region widths as part of your configuring, you load the region widths dialog. In this process then, should there be a cancel button there, or a back button (see first point). And should that button go back to the configure dialog or close?
- should we have *both* a cancel and a back button anywhere?

**Comment by stpaultim** [comment]
This feels like a bug and I don't think I saw it addressed above:

When I click on "Disable" for any layout, custom or standard, it just takes me from admin/structure/layouts/settings TO admin/structure/layouts/. Nothing else seems to happen. No error message. 

![Layout_settings___PR_2646_backdrop_backdrop_testing_site](https://user-images.githubusercontent.com/3144571/59982592-5ac85b80-95da-11e9-8a67-9feef3f5b46e.png)

**Comment by stpaultim** [comment]
By the way, I got my first real look at this today and am very impressed. Excellent work. 

One other comment. I was confused by the "0" shown below. It would appear that the default region names are not showing up. If a user edits and changes the region name from "content" to something else, then the new region name shows up. But "content" never shows up, instead we just see "0."

![Tim_s_Layout___PR_2646_backdrop_backdrop_testing_site](https://user-images.githubusercontent.com/3144571/59982796-302bd200-95dd-11e9-9fe7-b9ee074e0adb.png)

**Comment by docwilmot** [comment]
@stpaultim thanks for checking. The bulk operations doesnt work, youre right. The table was a "radical" idea at the time, so I just put something together to start and forgot about it. Most things on that table probably dont work.

Right now in a holding pattern till @jenlampton and @klonos clarify the recommendations above before I finish this off.

Thanks.

**Comment by jenlampton** [comment]
> Default drop-button action for flexible layout should be `Configure regions`

>> Not sure I agree; the default for other templates will be enable/disable, so why be different?

Drop-buttons are intended to be built so that the most-commonly used item will appear in the button, with all the less-commonly used items buried in the drop-down. For layouts that re not flexible, the most commonly used item is also the only item: "disable" when enabled, and "enable" when disabled. For a flexible layout, the most commonly used item will be `Configure regions`.

A good example of this change-in-drop-button-action-priority can be found on the Views listing. For enabled views the most common action is `Configure` but for disabled views it's `Enable`.

**Comment by jenlampton** [comment]
> Use more intuitive text for Row width behavior options

> > Already discussed, not settled

We know what we is there is not good enough for release, so let's choose something else that is, even if we know it's not perfect yet. 

Why don't we start with these:
`Maximum width for each screen size (adds the 'container-fluid' row class)`
`Fixed width for each screen size (adds the 'container' row class)`
`Always full width of page (no row classes added)`

We can try it for a while and change it after we know how it feels.

**Comment by jenlampton** [comment]
> Style bulk operations on layout template list to match admin/content

The top of the `admin/content` page shows views exposed filters on a white background, with a gray border and rounded corners.

<img width="1198" alt="Screen Shot 2019-07-05 at 4 20 58 PM" src="https://user-images.githubusercontent.com/397895/60748276-fe9df980-9f40-11e9-89b1-f00e292999fd.png">

I think this was just a request for a little matching CSS on the `admin/structure/layouts` page.

**Comment by jenlampton** [comment]
> Style human & machine names like we do on the content types page `admin/structure/types`

Or "Gold Standard" for displaying a Human name + Machine is as it's done on the content types page: 
<img width="1198" alt="Screen Shot 2019-07-05 at 4 20 58 PM" src="https://user-images.githubusercontent.com/397895/60748311-67857180-9f41-11e9-881a-e58c8053ec3a.png">

We have [an issue to add a theme function for this](https://github.com/backdrop/backdrop-issues/issues/3402)  but it's not in yet...

For the time being, this request is to add the machine name in parens next to the human readable name, so on the http://2646.backdrop.backdrop.qa.backdropcms.org/admin/structure/layouts/settings page the name for the `Drops Layer` Layout template would be displayed `Drops Layer (drops_layer)`

**Comment by jenlampton** [comment]
> > Merge icon and name columns on layout template list

> Doable, but necessary? (least difficult point)

Necessary, no. Improvement, yes. 

We need to remove the column heading 'Icon', and if possible, squish that column so it's the appropriate width (so the Name is right next to it's icon). It's usually easiest to do both these things by moving the icon into the other table cell.

**Comment by jenlampton** [comment]
> Hide the Column style/widths field via #ajax

This looks like it has been done already -- or at least -- the need for it has gone away. The Region widths selector is really slick now. 👍 

<img width="659" alt="Screen Shot 2019-07-05 at 4 45 05 PM" src="https://user-images.githubusercontent.com/397895/60748621-9ea95200-9f44-11e9-9a94-481a9114f15e.png">

<img width="639" alt="Screen Shot 2019-07-05 at 4 46 29 PM" src="https://user-images.githubusercontent.com/397895/60748627-a36e0600-9f44-11e9-8cda-51daba43b857.png">

**Comment by jenlampton** [comment]
> Consider moving the Number of columns setting into the fieldset, as a description.

The text from the  `Selected region widths` form element could moved (as a `#description`) into the `Regions` fieldset, along with where the region names are defined. 

For example: 
<img width="587" alt="Screen Shot 2019-07-05 at 5 05 54 PM" src="https://user-images.githubusercontent.com/397895/60748884-2f812d00-9f47-11e9-9fe5-270762a5d7da.png">

**Comment by jenlampton** [comment]
> > Add Back buttons and Cancel links to multi-step dialogs

> When adding a new row, we start with region widths dialog, so there will need to be a cancel button on the region widths dialog ... then we move to the row config dialog. Should there be a cancel there, or a back? If cancel, should it reload the region widths dialog or just close the dialog?

A `Cancel` link should cancel the whole thing and close the dialog. We could add a `Back` button too, that would go back to the region widths dialog.

> When configuring an existing row, we start with the row config dialog. A cancel button should be there now, I'd say 

Sounds good.

> then if you want to change the region widths as part of your configuring, you [click on the link to] load the region widths dialog. In this process then, should there be a cancel button there, or a back button (see first point). 

Interesting. I think the problem here is that the button on this dialog reads "Continue" as though it were in a sequence. If it were to read "Save region widths" then both the Save button and the Cancel link could return the user to the row config dialog. No need for a back button here. 

> should we have both a cancel and a back button anywhere?

Maybe we should try it with only Cancel links and see how that feels before we add Back buttons?

**Comment by docwilmot** [comment]
Mostly done with the todos
> Consider moving the Selected region widths setting into the Regions fieldset, as a description.

I've left this off for now. I dont think it works well with the collapsible fieldsets.
> Decide on language for more intuitive text for Row width behavior options

I've trimmed the text a bit, since it was too long and the resulting select element was extending outside of the fieldset and dialog width.

**Comment by stpaultim** [comment]
@docwilmot - Are you aware that the latest sandbox is not working. I'm getting errors when I visit the settings page - admin/structure/layouts/settings

**Comment by docwilmot** [comment]
I suspect I saved a dpm() in there somewhere. Out getting some exercise, will check later.

**Comment by docwilmot** [comment]
The settings table is working now. 
But I've removed the operations fieldset in favor of a submit button as I dont think operations would work in this context. We cant add any other actions to operations apart from enable and disable because those are the only actions that would be applicable to any group of templates. For example you cant select all the do "Configure rows" or "Delete", so in the end it would do excatly the same thing as a submit button. Except with an extra click

**Comment by docwilmot** [comment]
I think I've got all important the code written now. Now I'll do cleanup and move the flexible `template.inc` code to `admin.inc` where it belongs, and then write tests. Got 2 months to get that done! I think this might make it.

**Comment by docwilmot** [comment]
Last commit was more of a major renovation than a cleanup. Sandbox will need rebooting.

I'm thinking though before merging everything into `admin.inc` would be good to have a "code concept review" from @quicksketch. This PR, I'm just thinking, sort of abuses the Layout class in some ways, and would be good to get the concept vetted, and maybe ideas of any other way it should could be done.

**Comment by stpaultim** [comment]
We have layouts that use the descriptor 'hero' region to allow any region to be made full width while allowing other regions to be fixed-width. 

I believe that you have that option for single column regions in the flexible layout. Would it be helpful to use a term like 'hero region' or 'full screen width' to make that distinction. Describing the functionality with container classes is not helpful to most users. Sorry, the sandbox is broken so I can't login to be more specific.

**Comment by quicksketch** [comment]
I gave this a basic code and concept review. LOOKS AWESOME! :tada: 

The UI still has some work to do and I added a few more items to the TODO list in the issue summary.

There are a number of new comments I posted to https://github.com/backdrop/backdrop/pull/2646

> This PR, I'm just thinking, sort of abuses the Layout class in some ways, and would be good to get the concept vetted

The new `LayoutRendererFlexible` class seems a little magical to me, and the way that you're editing a layout template by essentially editing an empty Layout object is a little unexpected. However, I don't think this is a blocker. It sure seems like it has made things a lot easier by taking this approach.

There are a lot of UI issues that are still being worked through, and I didn't focus too much on that since I think this is on the right track.

The only terminology issue I had was the "Regions" vs "Columns" decision. When describing this verbally, I think I would say "columns" because that has immediate meaning in relation to "rows". "Regions per row" doesn't make nearly as much sense as "Columns per row". In other layouts, the word "region" is used because it's not necessarily locked to a rows/columns paradigm. But in this layout builder, we're only talking about columns within rows.

**Comment by klonos** [comment]
Not too fussed about using "columns" vs "regions", but once the viewport shifts to narrow/mobile screens, then I believe that the notion of columns is lost, as most of them are in that case stacked vertically on top of each other. Taking that into account, perhaps it does make more sense to call them "regions" instead of "columns". This reminds me of the situation of left/right sidebars, which got renamed to first/last (which is RTL/LTR-friendly).

**Comment by docwilmot** [comment]
> The new LayoutRendererFlexible class seems a little magical to me

Too magical. I've changed it to a standalone plugin (not extending Standard renderer). I think it makes more sense this way.
> Not too fussed about using "columns" vs "regions"

Me neither, and I prefer regions, because in this context they are regions. Adding "columns" as a concept will then have us disambiguating between column vs region to new users one day I'm sure.

I've done all the other changes and checkboxes, pushed to a new sandbox, and all is working. Will need to go through now and clean up the code and write tests. @klonos and @stpaultim would be grateful if you can start picking this apart meanwhile.

**Comment by stpaultim** [comment]
One thing I noticed that might not be easy to fix and I don't think it should be a blocker is that if you create a row, drag the row to a new position, then configure the row again, it looses it's position and drops back to the bottom. 

I don't think that is entirely unexpected, because the user is making multiple steps without saving their work. But, it would be great if the site remembered the location of the row before it was edited/configured and left it there afterwards, through the process of saving the layout.

**Comment by klonos** [comment]
1. click the "Configure row" button of any row
2. click "Change region widths" link
3. click "Cancel" -> the entire dialog closes

This is not an issue if you actually meant to change widths, but then changed your mind; but if you made changes between steps 1 and 2, these changes are lost if you hit the "Cancel" button in the "Choose region widths" dialog.

I would expect the either "Cancel" button to take me back to the "Configure row content" dialog, or a "back" button/link to take me to it. If I want to abort the entire editing at any point, I could use the x button at the top-right of either one of the dialogs.

Other nits:

- Why are we using textual representation of the widths in the "Selected region widths" field of the "Configure row content" dialog? Wouldn't it be better to have the visual ones instead?

- When clicking "Choose region widths", it always defaults to "1 100% region". I would expect it to default to whatever the row being edited is currently configured to.

**Comment by klonos** [comment]
> ...if you create a row, drag the row to a new position, then configure the row again, it looses it's position and drops back to the bottom.

Actually this happens if you configure any row - not the freshly-added and dragged one.

**Comment by klonos** [comment]
I had a (mostly) working copy of @docwilmot's PR where the icons were being automatically generated upon saving the template, to match the actual regions in the template. Not sure what I've done with it. ...can be a follow-up improvement.

**Comment by klonos** [comment]
I have added this issue to the "Layout improvements" project, and also in the "D8 feature parity" project, since the D8 layout builder already offers this functionality (although integrated in their Layout Builder, as opposed to a separate UI).

Also, during the latest dev meeting, we had a discussion about calling these things "regions" vs "columns". @quicksketch seems to be leaning towards "columns", because the things they are being added to are called "rows". Others (including myself) believe that we should call them "regions", mainly because that's what they end up being in the resulting template. Besides, if you view the template in narrow screens, these regions are being stacked on top of each other, which makes calling them "columns" odd.

During the same discussion, I proposed that we change the name "rows" to "sections", in order to avoid people correlating them with tables terminology ("rows", "columns"). It turns out that "sections" is what they are also called in the D8 Layout builder:

![Screen Shot 2019-08-24 at 12 54 43 am](https://user-images.githubusercontent.com/2423362/63625933-e2960a80-c609-11e9-8861-4c852be1463e.png)

**Comment by klonos** [comment]
...I have started a bikeshed/brainstorm on Gitter about an alternative name to "Flexible layouts":

@klonos:
> I really believe that calling the new customizable templates "flexible" calls for ambiguations with other things in the future. I have spent some time once again googling for synonyms of "custom" or "configurable" but nothing good came up. ...if only "LEGO" was not trademarked!
> ...does anyone else have any ideas?
> ...the only thing I have is "assemblates" 😅 

@herbdool:
> How do you see "flexible" as a problem?
> How about "layout builder"

@klonos:
> You mean "template builder". Yes, that what we can call the UI, but I'm talking about what we should call the templates themselves...
> ..."layout builder" is what the "Manage blocks" UI does.
> RE the problem I foresee: "Flexible" in web design means that the page adapts to various screen widths/sizes. Having templates specifically called "flexible" implies that the other, "standard" ones are not. Which is not true.

@klonos:
> The actual practical differences between the 2 types of templates are:
> - the ones provided as a layout project are a set of php files, while the "flexible" ones are config
> - the "standard" ones are read-only, whereas the "flexible" ones can be edited/changed

Other suggestions/ideas over "flexible":

@BWPanda: "sexy", "cubic"
@VasasA: "puzzle"
@klonos: "hopscotch", "tiler", "docwilmot templates"
@herbdool: "bricklayer", "configurable layout template"
@olafgrabienski: plain "layout template" (as in "Add layout template")
@laryn: "DIY templates", "Building Block Templates", "Combination Templates"
@findlabnet: "Mosaic"

...it is worth mentioning that while we were bikeshading on the name, @serundeputy was eating popcorn, watching us name things 😆

**Comment by klonos** [comment]
...I was re-reading #2119 (to figure out if we are actually solving that problem by the feature we are implementing here), and saw this comment by @quicksketch:

> ...I think what's being requested here is something that would just allow you to build your own dynamic layout. ... You'd assemble one layout and then it would be shown as an option side-by-side the hard-coded ones.

So perhaps another possible alternative would be "dynamic templates"? (not really fond of it - just trowing it out there).

**Comment by docwilmot** [comment]
> create a row, drag the row to a new position, then configure the row again, it looses it's position and drops back to the bottom.

I believe this is the exact same behavior as Layout blocks. Drag a block, configure it, it returns to the bottom. This could be a f/u issue then.
> I would expect either the "Cancel" button to take me back to the "Configure row content" dialog, or a "back" button/link to take me to it. If I want to abort the entire editing at any point, I could use the x button at the top-right of either one of the dialogs.

We had this discussion earlier in this thread. This is a reasonable catch though, will look into it.
> Why are we using textual representation of the widths in the "Selected region widths" field of the "Configure row content" dialog? Wouldn't it be better to have the visual ones instead?

Will check
> There is no drag indicator in the rows, so no clue that their order can be changed (unless you already know).

I tried to figure this earlier, but I couldnt do it. Would have to put the drag indicator outside the region marker dotted line, but I cant figure how to do that and keep the rows 100% at the same time. This may need a f/u by a design mind.
> Should we have a default icon that is not Simmons? Will people think they're getting exactly that layout instead of something that could be whatever they created?

I am no good at icon design, someone feel free to bang one together, we can always update it in a f/u issue.

Altogether though, none of these are blockers I'd say. Unless we disagree, can we still get a LGTM so that this PR falls into @quicksketch queue sooner?

**Comment by stpaultim** [comment]
What's the standard for LGTM (research tells me LGTM = looks good to me)?

It **Looks Good To Me** other than:
1) Some minor issues that can be follow-up issues.
2) It still says "Got some help here?" at top of admin/structure/layouts/settings/flexible-template/<template name>/configure 
I think help text here would be helpful, but is not required. We could remove that and create follow-up issue.
3) We kinda need to resolve the naming thing. 

I think that configurable layout templates is the most descriptive and accurate title discussed. But, it's long and clunky. I tried shortening configurable layout into 'conlay' or 'con-lay', but I'm not sure that works/is helpful. 

I like the sound of flexible better, but I understand the concerns of @klonos. Someone is going to have to make a choice. 

My suggestion is to go with configurable layout templates for now. I assume (?) that we could tweak this feature and change the name if we come up with something better anytime before actual release. 

I will be fine with whatever is decided (I'll just grumble quietly into my beer later). ;-)

**Comment by olafgrabienski** [comment]
Oh, when I delete a template in the PR, I get an "unexpected error":

- Deleted a layout template.
- Went to the corresponding layout.
- Unexpected error
  DBlog says: LayoutMissingException: The layout plugin "freaky" could not been found to display the "Freakshow" layout. in LayoutRendererStandard->__construct() (line 153 of /home/qa/github/backdrop/backdrop/pr/2646/core/modules/layout/plugins/renderers/layout_renderer_standard.inc)

**Comment by olafgrabienski** [comment]
>> "Flexible" in web design means that the page adapts to various screen widths/sizes. Having templates specifically called "flexible", implies that the other, "standard" ones are not; which is not true.

> I like the sound of flexible better, but I understand the concerns of @klonos.

Hm, I guess the term "responsive" is more common for screen size adaption than "flexible" but not sure.

However, in case we change the name let's remind that `flexible` isn't only part of the user interface it appears also in the code.

**Comment by klonos** [comment]
I think we can get this as a MVP, and improve iteratively over 1.14.x bug fix releases. I'll take some time to create follow-up issues for whatever can be done later on.

> Oh, when I delete a template in the PR, I get an "unexpected error"

Yeah, I have actually thought about testing that scenario myself, but I forgot: what happens to layouts and blocks when the layout template they are using is being disabled/removed? This sounds like a blocker, unless it is a pre-existing issue with "regular" layout templates.

> ...`flexible` isn't only part of the user interface, it appears also in the code.

Yup, that's why I am calling for a naming decision now, before we get this in. It will be really hard to change this once it's done.

> I think that configurable layout templates is the most descriptive and accurate title discussed. But, it's long and clunky.

Agreed on both points 😅 😓

...here's another suggestion: we could call them CLTs, short for "**C**onfigurable **L**ayout **T**emplates" ...same as [CCK](https://www.drupal.org/project/cck) for "Content Construction Kit", [EVA](https://www.drupal.org/project/eva) for "Entity Views Attachment", and [EMW](https://www.drupal.org/docs/7/api/entity-api/entity-metadata-wrappers) for "Entity Metadata Wrappers").

**Comment by klonos** [comment]
> This is difficult....

Don't stress about it. It can be a UX follow-up.

**Comment by klonos** [comment]
...maybe I'll wait before I file follow-ups 😄

**Comment by klonos** [comment]
Thanks @docwilmot 👍 

- When trying to disable templates in the PR sandbox, I get errors. Sometimes it may be a blank page (only the admin menu gets rendered), with the text `403` at the top-left; other times I get `The requested page "/admin/structure/layouts/settings/flexible-template/boxton/disable?token=ud7rgMARdKuhvXOA84rNdie2-hIbw6U8PvMx5bA5JLs" could not be found.`

- Can we please add a help text at the top of the list in `/admin/structure/layouts/settings`, along the lines of:

    > Disabled templates cannot be selected when creating/editing layouts. Templates that are already in use, cannot be disabled or deleted.

...I have left some additional feedback in the PR.

**Comment by klonos** [comment]
> When trying to disable templates in the PR sandbox, I get errors.

...I have tried clearing caches, but maybe it is time for a sandbox refresh.

**Comment by olafgrabienski** [comment]
Regarding an icon, what about a tetris style one, roughly like this?

![layout-flexible-icon](https://user-images.githubusercontent.com/14188307/63844599-c6d39100-c988-11e9-8647-aad40448b191.png)


Inspired by https://thenounproject.com/search/?q=tetris&i=721952#

See also:

- https://thenounproject.com/term/blocks

**Comment by klonos** [comment]
![ezgif com-gif-maker](https://user-images.githubusercontent.com/2423362/63891412-cb7e6080-c9ed-11e9-9ddf-140a03aa7a4b.png)

**Comment by klonos** [comment]
@docwilmot during todays' dev meeting, we have all agreed that "flexible" is the name to go with, so all good on that front 👍 🙂 ...now all that's left is to pick an icon and go with it! :shipit:

**Comment by quicksketch** [comment]
The icon is also a low impact change we can swap out at any time in a bugfix point release.

**Comment by quicksketch** [comment]
I re-reviewed https://github.com/backdrop/backdrop/pull/2646 and pushed a commit to that PR that fixed code style in several places and fixed an issue with selecting the first radio button always when changing between region (column) counts in a row.

However, the current pull request has issues when saving values.

- The "Save configuration" button when changing a row's settings does not save or close the dialog. This happens on editing existing rows or when adding a new one.
- Clicking "Configure row", then "Change region widths", select a different option, then click "Cancel". The new selected region style should be ignored, but instead it's saved, in the same way as though the "Save region widths" button was clicked.

**Comment by quicksketch** [comment]
Ah, I caused the first problem, and now fixed it.

The second problem where the "Cancel" button saves the layout regions when it shouldn't is still an issue, but not a complete show stopper. I'll see if I can work out that last issue.

**Comment by quicksketch** [comment]
Okay I think this is ready for another set of eyes again. Overall I made the following changes:

- I fixed the "Cancel" button problem described above.
- There was some code to "lock" and "break lock" layout templates as they were being edited. However, this code did not seem to actively be used and the break-lock callback had several undefined variables. As editing layout templates seems less prone to conflicts than layout configurations, I didn't think this was a big loss, so instead of attempting to fix it I just removed it.
- Enabling and disabling non-flexible layouts via the Enable/Disable button did not work. I moved this code into `layout.admin.inc` and made it work for for flexible and non-flexible layouts.
- Small function naming changes. I tried to increase consistency but naming is sort of all over the place.
- I made this icon for now. Not at all attached to it if we want to change it later:

![flexible_template](https://user-images.githubusercontent.com/100206/64135421-240e7e80-cd9d-11e9-8376-a8e575bc4fce.png)

**Comment by quicksketch** [comment]
One more set of changes based on @herbdool's review is up.

I was also bothered by the direct writing of config files in form submit handlers: when we have data structures, we should have official APIs for writing config. So I converted the `$flexible_template` variable everywhere from an array to an new `LayoutFlexibleTemplate` object. Now just like the `Layout` class, if you want to save or delete a template you can use `$flexible_template->save()` or `$flexible_template->delete()`.

I also change the pseudo-constructor to be a real one, now that we have an object.

Before
`$flexible_template = layout_flexible_template_template_default()` // Creates an array.

After
`$flexible_template = new LayoutFlexibleTemplate()` // Creates an object.

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/2646 into 1.x for 1.14.0. Thanks @BWPanda, @herbdool, @jenlampton, @klonos, @olafgrabienski, @stpaultim!! And a HUGE thanks to @docwilmot for spearheading this one!

**Comment by quicksketch** [comment]
@docwilmot Found a bug in our token checking for deleting rows. Follow-up PR at https://github.com/backdrop/backdrop/pull/2842

**Comment by quicksketch** [comment]
I got a spot check from @docwilmot in Gitter. Merged https://github.com/backdrop/backdrop/pull/2842 in with his feedback.
